### PR TITLE
Simple c++ base class for processing algorithms

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR introduces a new c++ base class for processing algorithms. It's another step along the way in porting the processing framework to c++.

So far it's quite basic - it just handles groups/names/flags/icons. There's no logic here yet. But I need this part merged and a c++ algorithm class in order to complete the port of the providers and alglist to c++.

### CODE REVIEW

File: src/core/processing/qgsprocessingalgorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "qgsprocessingalgorithm.h"
+#include "qgsapplication.h"
+
+QIcon QgsProcessingAlgorithm::icon() const
+{
+  return QgsApplication::getThemeIcon( "/processingAlgorithm.svg" );
+}
+
+QString QgsProcessingAlgorithm::svgIconPath() const
```
Comment (m-kuhn) [inline src/core/processing/qgsprocessingalgorithm.cpp]: Any reason to explicitly insist on svg here?

File: src/core/processing/qgsprocessingalgorithm.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+{
+  public:
+
+    //! Flags indicating how and when an algorithm operates and should be exposed to users
+    enum Flag
+    {
+      FlagHideFromToolbox = 1 << 1, //!< Algorithm should be hidden from the toolbox
+      FlagHideFromModeler = 1 << 2, //!< Algorithm should be hidden from the modeler
+      FlagSupportsBatch = 1 << 3,  //!< Algorithm supports batch mode
+      FlagDeprecated = FlagHideFromToolbox | FlagHideFromModeler, //!< Algorithm is deprecated
```
Comment (m-kuhn) [inline src/core/processing/qgsprocessingalgorithm.h]: Nice idea to put it this way.

**Comment by m-kuhn** [comment]
Yes please

File: src/core/processing/qgsprocessingalgorithm.cpp (hunk 1 above)
Comment (nyalldawson) [inline src/core/processing/qgsprocessingalgorithm.cpp]: It's a request from @nirvn, used for high resolution exports of the model

**Comment by nyalldawson** [comment]
@volaya @alexbruy any thoughts? I'd like to merge this as more followup changes are inbound shortly :)

**Comment by alexbruy** [comment]
Looks good for me

**Review by alexbruy — APPROVED** [review]

**Pull request merged**